File: call_manager/app/src/main/java/com/designated/callmanager/data/DriverInfo.kt
package com.designated.callmanager.data

import com.designated.callmanager.data.Constants
import com.google.firebase.Timestamp
import com.google.firebase.firestore.PropertyName

/**
 * Firestore의 기사 정보를 나타내는 데이터 클래스 (Firestore 구조 기준)
 */
data class DriverInfo(
    @PropertyName("id") var id: String = "", // Firestore 문서 ID
    @PropertyName("authUid") var authUid: String? = null, // Firestore 'authUid' 필드 매핑 (로그 기준)
    @PropertyName("name") var name: String = "",
    @PropertyName("status") var status: String = Constants.DRIVER_STATUS_OFFLINE, // enum 대신 상수 사용
    @PropertyName("phoneNumber") var phoneNumber: String = "",
    @PropertyName("createdAt") var createdAt: Timestamp? = null, // Nullable로 변경 (Firestore에 없을 경우 대비)
    @PropertyName("updatedAt") var updatedAt: Timestamp? = null, // Firestore 'updatedAt' 필드 매핑 (updateAt -> updatedAt)
    @PropertyName("regionId") var regionId: String? = null, // Firestore 'regionId' 필드 매핑 (regionID -> regionId)
    @PropertyName("officeId") var officeId: String? = null, // Firestore 'officeId' 필드 매핑 (officeID -> officeId)
    @PropertyName("email") var email: String? = null,
    @PropertyName("driverType") var driverType: String? = null,
    @PropertyName("approvedAt") var approvedAt: Timestamp? = null,
    @PropertyName("approvalStatus") var approvalStatus: String = Constants.APPROVAL_STATUS_PENDING // String 타입 및 상수 사용
) {
    // Firestore에서 객체 매핑 시 빈 생성자 필요 - 모든 필드에 기본값을 제공하여 명시적 빈 생성자의 복잡성을 줄임
    constructor() : this(
        id = "",
        authUid = null,
        name = "",
        status = Constants.DRIVER_STATUS_OFFLINE, // enum 대신 상수 사용
        phoneNumber = "",
        createdAt = null,
        updatedAt = null,
        regionId = null,
        officeId = null,
        email = null,
        driverType = null,
        approvedAt = null,
        approvalStatus = Constants.APPROVAL_STATUS_PENDING // String 타입 및 상수 사용
    )
}

File: driver_app_flutter/lib/models/driver_info.dart
import 'package:cloud_firestore/cloud_firestore.dart';

enum DriverStatus {
  online('ONLINE', '대기중'),
  offline('OFFLINE', '오프라인'),
  busy('BUSY', '운행중'),
  waiting('WAITING', '배차대기');

  final String value;
  final String displayName;

  const DriverStatus(this.value, this.displayName);

  static DriverStatus fromString(String? status) {
    if (status == null) return offline;
    return DriverStatus.values.firstWhere(
      (e) => e.value == status.toUpperCase(),
      orElse: () => offline,
    );
  }
}

enum DriverApprovalStatus {
  pending('PENDING', '승인대기'),
  approved('APPROVED', '승인완료'),
  rejected('REJECTED', '거절됨');

  final String value;
  final String displayName;

  const DriverApprovalStatus(this.value, this.displayName);

  static DriverApprovalStatus fromString(String? status) {
    if (status == null) return pending;
    return DriverApprovalStatus.values.firstWhere(
      (e) => e.value == status.toUpperCase(),
      orElse: () => pending,
    );
  }
}

class DriverInfo {
  final String id;
  final String authUid;
  final String name;
  final String phone;
  final String regionId;
  final String officeId;
  final DriverStatus status;
  final DriverApprovalStatus approvalStatus;
  final String? fcmToken;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  DriverInfo({
    required this.id,
    required this.authUid,
    required this.name,
    required this.phone,
    required this.regionId,
    required this.officeId,
    required this.status,
    required this.approvalStatus,
    this.fcmToken,
    this.createdAt,
    this.updatedAt,
  });

  /// Firestore에서 복원
  factory DriverInfo.fromFirestore(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>;
    return DriverInfo(
      id: doc.id,
      authUid: data['authUid'] as String,
      name: data['name'] as String,
      phone: data['phone'] as String,
      regionId: data['regionId'] as String,
      officeId: data['officeId'] as String,
      status: DriverStatus.fromString(data['status'] as String?),
      approvalStatus: DriverApprovalStatus.fromString(data['approvalStatus'] as String?),
      fcmToken: data['fcmToken'] as String?,
      createdAt: (data['createdAt'] as Timestamp?)?.toDate(),
      updatedAt: (data['updatedAt'] as Timestamp?)?.toDate(),
    );
  }

  /// Firestore에 저장할 Map
  Map<String, dynamic> toMap() {
    return {
      'authUid': authUid,
      'name': name,
      'phone': phone,
      'regionId': regionId,
      'officeId': officeId,
      'status': status.value,
      'approvalStatus': approvalStatus.value,
      'fcmToken': fcmToken,
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }

  /// Copy with
  DriverInfo copyWith({
    String? id,
    String? authUid,
    String? name,
    String? phone,
    String? regionId,
    String? officeId,
    DriverStatus? status,
    DriverApprovalStatus? approvalStatus,
    String? fcmToken,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return DriverInfo(
      id: id ?? this.id,
      authUid: authUid ?? this.authUid,
      name: name ?? this.name,
      phone: phone ?? this.phone,
      regionId: regionId ?? this.regionId,
      officeId: officeId ?? this.officeId,
      status: status ?? this.status,
      approvalStatus: approvalStatus ?? this.approvalStatus,
      fcmToken: fcmToken ?? this.fcmToken,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
